src/daft-core/src/array/ops/date.rs

```diff
@@ -195,4 +219,93 @@ impl TimestampArray {
             Box::new(date_arrow),
         )
     }
+
+    pub fn truncate(&self, interval: &str, start_time: &Option<i64>) -> DaftResult<Self> {
+        let physical = self.physical.as_arrow();
+        let DataType::Timestamp(timeunit, tz) = self.data_type() else {
+            unreachable!("Timestamp array must have Timestamp datatype")
+        };
+        let tu = timeunit.to_arrow();
+        let duration = process_interval(interval)?;
+
+        macro_rules! truncate_dt {
+            ($dt:expr, $duration:expr) => {
+                match $dt.duration_trunc($duration) {
+                    Ok(dt) => match tu {
+                        arrow2::datatypes::TimeUnit::Second => Ok(dt.timestamp()),
+                        arrow2::datatypes::TimeUnit::Millisecond => Ok(dt.timestamp_millis()),
+                        arrow2::datatypes::TimeUnit::Microsecond => Ok(dt.timestamp_micros()),
+                        arrow2::datatypes::TimeUnit::Nanosecond => {
+                            dt.timestamp_nanos_opt()
+                                .ok_or(DaftError::ValueError(format!(
+                                    "Error truncating timestamp, result is out of range : {{dt}}"
+                                )))
+                        }
+                    },
+                    Err(RoundingError::DurationExceedsTimestamp) => Ok(0),
+                    Err(e) => Err(DaftError::ValueError(format!(
+                        "Error truncating timestamp: {e}"
+                    ))),
+                }
+            };
+        }
+
+        let result_timestamps = physical
+            .iter()
+            .map(|ts| {
+                ts.map_or(Ok(None), |ts| {
+                    let adjusted_ts = match start_time {
+                        Some(st) if *st > *ts => {
+                            return Err(DaftError::ValueError(format!(
+                                "Start time is greater than timestamp: {st} > {ts}"
```

Comment:
I wonder if this maybe is ok? Could we allow for negative truncation (e.g. `-1 hour`) relative to start time?

Then start time is less so "start_time" and more so "relative_to_time" or something